libp2p/autotls/peeridauth.nim

```diff
@@ -0,0 +1,202 @@
+
+proc encodeVarint(n: int): seq[byte] =
+  var varInt: seq[byte] = @[]
+  var x = uint64(n)
+  while x != 0:
+    var byteVal = byte(x and 0x7F)
+    x = x shr 7
+    if x != 0:
+      byteVal = byteVal or 0x80
+    varInt.add(byteVal)
+  return varInt
+
+proc genDataToSign(prefix: string, parts: seq[SigParam]): seq[byte] =
+  var buf: seq[byte] = prefix.toByteSeq()
+  for p in parts:
+    buf.add encodeVarint(p.k.len + p.v.len + 1)
+    buf.add (p.k & "=").toByteSeq
+    buf.add p.v
+  return buf
+
+proc peerIdSign(
+    privateKey: PrivateKey,
+    challenge: string,
+    publicKey: PublicKey,
+    hostname: string,
+    clientSender: bool = true,
+): string =
+  let prefix = "libp2p-PeerID"
+  let parts =
+    # parts need to be sorted alphabetically by key
+    if clientSender:
+      @[
+        SigParam(k: "challenge-client", v: challenge.toByteSeq()),
+        SigParam(k: "hostname", v: hostname.toByteSeq()),
+        SigParam(k: "server-public-key", v: publicKey.getBytes().get()),
+      ]
+    else:
+      @[
+        SigParam(k: "challenge-server", v: challenge.toByteSeq()),
+        SigParam(k: "client-public-key", v: publicKey.getBytes().get()),
+        SigParam(k: "hostname", v: hostname.toByteSeq),
+      ]
+  let bytesToSign = genDataToSign(prefix, parts)
+  return base64.encode(privateKey.sign(bytesToSign).get().getBytes(), safe = true)
+
+proc checkSignature(
+    serverSig: string,
+    serverPublicKey: PublicKey,
+    challengeServer: string,
+    clientPublicKey: PublicKey,
+    hostname: string,
+): bool {.raises: [PeerIDAuthError].} =
+  let prefix = "libp2p-PeerID"
+  let parts: seq[SigParam] =
+    @[
+      SigParam(k: "challenge-server", v: challengeServer.toByteSeq()),
+      SigParam(k: "client-public-key", v: clientPublicKey.getBytes().get()),
+      SigParam(k: "hostname", v: hostname.toByteSeq),
+    ]
+  let signedBytes = genDataToSign(prefix, parts)
+  var serverSignature: Signature
+  try:
+    if not serverSignature.init(base64.decode(serverSig).toByteSeq):
+      raise newException(
+        PeerIDAuthError, "Failed to initialize Signature from base64 encoded sig"
+      )
+  except ValueError:
+    raise newException(PeerIDAuthError, "Failed to decode server's signature")
+  return serverSignature.verify(
+    signedBytes.toOpenArray(0, signedBytes.len - 1), serverPublicKey
+  )
+
+proc peerIdAuthenticate(
+    url: string, session: HttpSessionRef, peerInfo: PeerInfo, payload: JsonNode
+): Future[(string, HttpClientResponseRef)] {.
+    async: (raises: [AutoTLSError, PeerIDAuthError, CancelledError])
+.} =
+  # Authenticate in three ways as per the PeerID Auth spec
+  # https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md
+
+  # request authentication
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  var authStartResponse: HttpClientResponseRef
+  try:
+    authStartResponse = await HttpClientRequestRef.get(session, url).get().send()
```

Comment:
Under which conditions could HttpClientRequestRef.get(session, url).get() return an error? This assumes it wont, right?